sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java

```diff
@@ -17,96 +17,120 @@
+      BUILTIN_AGGREGATOR_FACTORIES =
+          ImmutableMap.<String, Function<Schema.TypeName, CombineFn<?, ?, ?>>>builder()
+              .put("COUNT", typeName -> Count.combineFn())
+              .put("MAX", BeamBuiltinAggregations::createMax)
+              .put("MIN", BeamBuiltinAggregations::createMin)
+              .put("SUM", BeamBuiltinAggregations::createSum)
+              .put("$SUM0", BeamBuiltinAggregations::createSum)
+              .put("AVG", BeamBuiltinAggregations::createAvg)
+              .put("VAR_POP", VarianceFn::newPopulation)
+              .put("VAR_SAMP", VarianceFn::newSample)
+              .put("COVAR_POP", CovarianceFn::newPopulation)
+              .put("COVAR_SAMP", CovarianceFn::newSample)
+              .build();
+
   private static MathContext mc = new MathContext(10, RoundingMode.HALF_UP);
 
+  public static CombineFn<?, ?, ?> create(String functionName, Schema.TypeName fieldTypeName) {
+
+    Function<Schema.TypeName, CombineFn<?, ?, ?>> aggregatorFactory =
+        BUILTIN_AGGREGATOR_FACTORIES.get(functionName);
+
+    if (aggregatorFactory != null) {
+      return aggregatorFactory.apply(fieldTypeName);
+    }
+
+    throw new UnsupportedOperationException(
+        String.format("Aggregator [%s] is not supported", functionName));
+  }
+
   /** {@link CombineFn} for MAX based on {@link Max} and {@link Combine.BinaryCombineFn}. */
-  public static CombineFn createMax(SqlTypeName fieldType) {
+  static CombineFn createMax(Schema.TypeName fieldType) {
     switch (fieldType) {
       case BOOLEAN:
-        return new CustMax<Boolean>();
-      case INTEGER:
+      case INT16:
+      case BYTE:
+      case FLOAT:
+      case DATETIME:
+      case DECIMAL:
+        return new CustMax<>();
+      case INT32:
         return Max.ofIntegers();
-      case SMALLINT:
-        return new CustMax<Short>();
-      case TINYINT:
-        return new CustMax<Byte>();
-      case BIGINT:
+      case INT64:
         return Max.ofLongs();
-      case FLOAT:
-        return new CustMax<Float>();
       case DOUBLE:
         return Max.ofDoubles();
-      case TIMESTAMP:
-        return new CustMax<ReadableInstant>();
-      case DECIMAL:
-        return new CustMax<BigDecimal>();
       default:
         throw new UnsupportedOperationException(
             String.format("[%s] is not support in MAX", fieldType));
     }
   }
 
   /** {@link CombineFn} for MIN based on {@link Min} and {@link Combine.BinaryCombineFn}. */
-  public static CombineFn createMin(SqlTypeName fieldType) {
+  static CombineFn createMin(Schema.TypeName fieldType) {
     switch (fieldType) {
       case BOOLEAN:
-        return new CustMin<Boolean>();
-      case INTEGER:
+      case BYTE:
+      case INT16:
+      case FLOAT:
+      case DATETIME:
+      case DECIMAL:
+        return new CustMin();
+      case INT32:
         return Min.ofIntegers();
-      case SMALLINT:
```

Comment:
Is this fixing a bug that Schema has no longer SMALLINT type (and other integer types)?

Can we create a JIRA to improve the test coverage on aggregation such that we could discover the breaking changes.

Reply:
It supports everything, I just changed it from Calcite types to Beam Schema types

Reply:
I see. Didn't aware that there is a conversion of types.

Reply:
There's no type conversion here, here we choose a specific implementation of the combiner based on the field type, e.g. `Sum.ofIntegers()` vs `Sum.ofLongs()` and similar. Calcite types are not serializable and as soon as you need to pass them around, things get complicated.

Reply:
Gotcha.